RFD.md

```diff
@@ -0,0 +1,217 @@
+---
+authors: Oleksiy Kovyrin (oleksiy@kovyrin.net)
+state: draft
+---
+
+# RFD - Teleport Systems Challenge
+
+## What
+
+This document will be used to communicate and discuss design decisions for the
+implementation of Teleport Systems Challenge by Oleksiy Kovyrin. See the [challenge page](https://github.com/gravitational/careers/blob/main/challenges/systems/challenge.md) for more details.
+
+## Why
+
+This exercise has two goals:
+
+* It helps the Teleport team understand what to expect from Oleksiy as a developer, how he writes production code, how he reasons about API design and how he communicates when trying to understand a problem before solving it.
+* It helps Oleksiy get a feel for what it would be like to work at Teleport, as this exercise aims to simulate their day-as-usual and expose Oleksiy to the type of work Teleport engineers are doing.
+
+## Requirements
+
+We need to build a client-server system that allows for remote execution of shell commands with per-command resource control (memory, CPU and disk IO) and isolation on network, process and filesystem levels using Linux kernel technologies like cgroups and kernel namespaces.
+
+Here are the original requirements as described on the [challenge documentation page](https://github.com/gravitational/careers/blob/main/challenges/systems/challenge.md):
+
+
+### Library
+
+* Worker library with methods to start/stop/query status and get the output of a job.
+* Library should be able to stream the output of a running job.
+  * Output should be from start of process execution.
+  * Multiple concurrent clients should be supported.
+* Add resource control for CPU, Memory and Disk IO per job using cgroups.
+* Add resource isolation for using PID, mount, and networking namespaces.
+
+### API
+
+* [GRPC](https://grpc.io) API to start/stop/get status/stream output of a running process.
+* Use mTLS authentication and verify client certificate. Set up strong set of cipher suites for TLS and good crypto setup for certificates. Do not use any other authentication protocols on top of mTLS.
+* Use a simple authorization scheme.
+
+### Client
+
+* CLI should be able to connect to worker service and start, stop, get status, and stream output of a job.
+
+## Solution Details
```

Comment:
Could you add a couple sections to solution details? In particular, I'm interested in more information on:

- the streaming solution
- the process lifecycle (how will the server/library handle starting a job and setting up the isolation)

Reply:
Added a separate section with the information requested. I will be honest: this is starting to get way too deep for a design document in my opinion since I'm forced to speculate on how some pretty low-level stuff will be implemented. I would prefer to leave those details for the actual implementation where I can experiment and see where things go instead of trying to nail them down in a document.